src/wal/mod.rs

```diff
@@ -129,3 +123,136 @@ impl Wal {
         Ok(())
     }
 }
+
+/// Open a WAL file and check the integrity of the file, it decodes
+/// all records and make sure they are coherent with the WAL format.
+pub struct WalChecker {
+    last_batch: Option<Batch>,
+}
+
+pub enum ConsistencyError {
+    LastBatchCrashed(Batch),
+    NotConsistent(u64),
+}
+
+impl WalChecker {
+    /// Open a WAL file and check its integrity, if a problem is found then the
+    /// wal is restored to its last valid state, thus it updates the WAL file
+    /// by removing everything after the last valid record.
+    ///
+    /// NOTE: Even though currently in the simulation each batch is pruned right after
+    /// the success of the next one, the following method needs to be able to handle multiple
+    /// records of different batches and validate them
+    pub fn open_and_recorver(path: PathBuf) -> Self {
+        let mut wal_file = match OpenOptions::new().read(true).write(true).open(&path) {
+            Ok(wal_file) => wal_file,
+            Err(err) if err.kind() == std::io::ErrorKind::NotFound => {
+                // a non-existing WAL is considered valid, with the assumed sequence number being zero
+                return Self { last_batch: None };
+            }
+            Err(_) => {
+                panic!("Error opening or creating the wal file ")
+            }
+        };
+
+        let (last_records, last_records_position) = Self::validate_records(&mut wal_file);
+
+        // If not all the WAL was parsed correctly, then the recovery procedure means to remove
+        // all incorrect records from the last valid one
+
+        let wal_file_len = wal_file
+            .metadata()
+            .expect("Error extracting wal file len")
+            .len();
+
+        if last_records_position as u64 != wal_file_len {
```

Comment:
Something is missing here, but maybe could just be seen as an optimization.

so this function returns the:
  1. last complete set of records within the WAL
  2. the end position of that complete set of records
  
note that there are two possibilities:
  1. this set of records is at the end of the file. in this case, it's possible that these records were incompletely applied to the DB and should be reapplied
  2. _or_ this set of records is not at the end of the file. in this case, we know that these records _were_ completely applied and all we need to do is truncate the WAL to `last_records_position`.

What I'm saying is that we _either_ need to truncate the WAL _or_ reapply the last batch, but never both. The code as written does both. I think adding an `else` here which sets `last_batch` only in the case that `last_records_position == wal_file_len`.

Reply:
> The code as written does both

I would say that the code as written **could** do both, but currently it does not (given how the simulation writes things into the wal and the store). Inside main.rs, there is a call to `check_consistency` where `last_batch` is kept only if there is an inconsistency with the store. Otherwise, like you said, if the wal would have been truncated, then for sure it will be consistent (and nothing will be re-applied)

Adding an else statement to the function, as you described, would reduce the little generality of the `WalChecker`, which allows for easy changes in the simulation for testing pipelining. In that case, there could possibly be a crash resulting in the Wal batch failing to be written, along with the previous set of pages fail being written to the store

Reply:
> The code as written does both

I would say that the code as written could do both, but currently it does not (given how the simulation writes things into the wal and the store). Inside main.rs, there is a call to check_consistency where last_batch is kept only if there is an inconsistency with the store. Otherwise, like you said, if the wal would have been truncated, then for sure it will be consistent (and nothing will be re-applied)

Adding an else statement to the function, as you described, would reduce the little generality of the WalChecker, which allows for easy changes in the simulation for testing pipelining. In that case, there could possibly be a crash resulting in the Wal batch failing to be written, along with the previous set of pages fail being written to the store